File: 42 - Generate All Possible IP Addresses/solution.js
function generateIPAddresses(str) {
  const result = [];

  function backtrack(start, path) {
    if (path.length === 4 && start === str.length) {
      result.push(path.join("."));
      return;
    }

    if (path.length === 4 || start >= str.length) {
      return;
    }

    for (let len = 1; len <= 3; len++) {
      const part = str.slice(start, start + len);
      if (part.length > 1 && part[0] === "0") continue;
      if (+part > 255) continue;
      backtrack(start + len, [...path, part]);
    }
  }

  backtrack(0, []);
  return result;
}

// Test cases

console.log(generateIPAddresses("25525511135"));

File: 42 - Generate All Possible IP Addresses/solution.py
def restore_ip_addresses(s):
    result = []

    def backtrack(start, path):
        if len(path) == 4:
            if start == len(s):
                result.append('.'.join(path))
            return

        for length in range(1, 4):  # Try segments of length 1, 2, or 3
            if start + length > len(s):
                break
            segment = s[start:start + length]
            
            # Skip if the segment has leading zeros or is > 255
            if (segment.startswith('0') and len(segment) > 1) or int(segment) > 255:
                continue

            backtrack(start + length, path + [segment])

    backtrack(0, [])
    return result


# Test cases
print(restore_ip_addresses('25525511135'))